feat: join built-in function

zoldater opened the pull request:
## Description

The pull request introduces the `$join()` built-in function that returns a string value concatenating array elements with separator.

## Motivation

`sprintf` function replaces existing specifiers with the values from `arguments_list `. For this reason passing `$group()`, `$team()`, `$assignees()` or any array of arbitrary size as `arguments_list` could be correct.

Proposed change makes it possible to join arrays into string to replace only one format specifier.

## Example

- `$sprintf("These changes require approval by team 'a': %s", [$join($group("a"))])`

## Related issue

<!-- Closes # (issue) -->

## Type of change

<!-- Uncomment the right types of change from the options bellow: -->

<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Improvements (non-breaking change without functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## How was this tested?

<!-- Please describe the tests that you ran to verify your changes. -->
Unit tests.
<!-- Provide instructions so we can reproduce (if applicable.) -->
<!-- Please also list any relevant details for your test configuration (if applicable.) -->

## Checklist

<!-- All checks are required in order to open a pull request ready to review. -->

- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works <!-- Delete this if not applicable -->
- [x] I have ran `task check -f` and have no issues

## Code review and merge strategy (ship/show/ask) 

<!-- Please uncomment and check only *one* of the following -->

<!-- - [ ] Ship: this pull request can be automatically merged and does not require code review --> 
<!-- - [ ] Show: this pull request can be auto-merged and code review should be done post merge --> 
- [x] Ask: this pull request requires a code review before merge

marcelosousa commented on engine/transform_internal_test.go:
```diff
@@ -115,6 +115,46 @@ func TestTransformAladinoExpression(t *testing.T) {
+		"join assignees with non-empty separator": {
+			arg:     `$join($assignees(), ", ")`,
+			wantVal: `$join($assignees(), ", ")`,
+		},
```
The transform has a limitation that it doesn't work for nested $join statements. For example:

```
        "nested join with empty separators ": {
			arg:     `$join([$join(["a","b"]), "c"])`,
			wantVal: `$join([$join(["a","b"], " "), "c"], " ")`,
		},
```

zoldater replied:
Thanks @marcelosousa !
I've added the test and fixed issue with recursion.

marcelosousa reviewed (commented):
Many thanks for this PR 🙌  I've left a comment related to the transform of the join built-in as it is not working for nested calls. I believe this is a broader issue of the way we are doing these transformations and should be addressed in a different issue. @ferreiratiago should I open an issue for that so that we can unblock this PR?

For the rest, it's looking good - thanks!

marcelosousa commented:
Test coverage is good!

zoldater commented:
Hi @ferreiratiago and @marcelosousa !
I fixed the issue with nested join calls and added proposed test to `engine/transform_internal_test.go`. So it should not be necessary to open new issue for that.

ferreiratiago commented on engine/transform.go:
```diff
@@ -73,6 +73,21 @@ func addDefaultHaveAllChecksRunCompleted(str string) string {
 }
 
+func addDefaultJoinSeparator(str string) string {
+	r := regexp.MustCompile(`\$join\(((\[.*\])|(\$(team|group)\("[^"]*"\))|(\$assignees\(\)))(,\s*"([^"]*)")?\)`)
```
What's the reasoning behind `team`, `group`, and `assignees`? I may be missing something 😅

Isn't a simpler regex enough?
```suggestion
	r := regexp.MustCompile(`\$join\((\[.*\]|([^,]*))(,\s*"([^"]*)")?\)`)
```

If we have the literal values `team`, `group`, and `assignees`, we need to keep updating this transform every time there's a new built-in that returns a []string. For instance, [$filePaths()](https://github.com/reviewpad/reviewpad/issues/538).

zoldater replied:
Thanks for pointing this out! Simplified the regex.

zoldater commented on plugins/aladino/functions/join.go:
```diff
@@ -0,0 +1,37 @@
+		switch v := val.(type) {
+		case *aladino.StringValue:
+			clearVals = append(clearVals, v.Val)
+		}
```
Done.

ferreiratiago commented on plugins/aladino/functions/join.go:
```diff
@@ -0,0 +1,40 @@
+		case *aladino.StringValue:
+			clearVals = append(clearVals, v.Val)
+		default:
+			return nil, fmt.Errorf("join: invalid element of kind %v", v.Kind())
```
We are very close. Just missing the unit test for this case.

zoldater replied:
Added `TestJoin_BadType`

ferreiratiago reviewed (changes requested):
One unit test missing. Other than that it's perfect 🙌

ferreiratiago reviewed (approved):
Perfect. Thanks for your contribution @zoldater 🙏
